Load the embedded animation and make the engine play it on demand

lhog opened the pull request:
EARLY DRAFT
<video src="https://github.com/user-attachments/assets/1dc9acda-5686-4262-9f5e-05c71e0a41a5" controls width="800"></video>

<video src="https://github.com/user-attachments/assets/af0c34a9-51f5-4c23-ab5e-0a90ab2cdc3e" controls width="800"></video>

lhog committed:
Load the embedded animation from GLTF

lhog committed:
Add embedded animation playback with keyframe sampling and Lua bindings

Animation sampling infrastructure (3DModelAnimation):
- Add SampleSequence<T> with LERP for float3/float and SLERP for quaternions
- Add FindKeyframeBracket binary search helper for time->keyframe lookup
- Add HasAnimation() and GetAnimationDuration() query helpers

Embedded animation player (UnitScript):
- Add EmbeddedAnimPlayer struct with creg serialization (name, time, speed, looping)
- TickEmbeddedAnim samples keyframes each tick and applies translation/rotation/scale
  to pieces, respecting script-anim channel ownership (Move/Turn/Spin/Scale take priority)
- Play/Stop control methods with duration lookup from the model's animation map
- HaveAnimations() now considers active embedded animations

Lua bindings (LuaUnitScript):
- Spring.PlayAnimation(name, speed, loop)
- Spring.StopAnimation()
- Spring.SetAnimationSpeed(speed)
- Spring.SetAnimationTime(time)
- Spring.GetAnimationTime() -> number
- Spring.GetAnimationDuration() -> number

lhog committed:
Fixes

lhog committed:
Expose embedded animation Lua bindings and query API

Add PlayAnimation/StopAnimation/SetAnimationSpeed/SetAnimationTime/
GetAnimationTime/GetAnimationDuration to the unit script header.
Add Spring.GetUnitDefAnimationNames to query animation names from a
unit def's model. Expose animation map iterators and size accessor on
ModelAnimation. Fix PlayAnimation loop default to false instead of true.

First tests are successful.

lhog committed:
Fix the crash and unreasonable animation speedups

sprunk commented:
* what happens if I manually call `Turn` etc on a piece that is undergoing an animation? Maybe it should be a no-op, or even raise an error?
* could multiple animations play in parallel if they affect non-overlapping piece sets? E.g. play movement animation on legs, some reload animation on weapon, and some "scanning" animation on a radar dish all at once? Looks like there's only one animation available at a time now.
* there doesn't seem to be an "is animation playing?" call. You can call one of the others and discard the result (just check for nil) but an explicit one would be nice.

lhog committed:
Add multi-track embedded animation blending with restore commands

lhog committed:
Make PlayEmbeddedAnimation reentrant, other minor changes

lhog committed:
Add PlayAnimation wait support for embedded animations

Adds a  parameter to Spring.UnitScript.PlayAnimation that blocks
the calling thread until a non-looping animation finishes. Introduces
a new AnimationFinished callin that Lua scripts can implement to be
notified when animations complete.

Changes:
- Add waitingForAnimation tracking in Lua gadget framework
- Add AnimationFinished callin to LuaScriptNames
- Add EmbeddedAnimFinished to CUnitScript for callback dispatch
- Update PlayEmbeddedAnimation to accept wait parameter
- Add hasWaiting/hasFiredCompletion tracking in EmbeddedAnimPlayer

lhog committed:
Fix CLuaUnitScript::StopAnimation()

lhog committed:
Bugfixes

lhog committed:
refactor: replace string-based animation lookups with ID-based indexing in unit scripts and GLTF parsing

lhog committed:
WIP

lhog committed:
Optimize embedded animation tick loop: cache channels, replace FindAnim scans with claim bitmap, use size_t for animation IDs
